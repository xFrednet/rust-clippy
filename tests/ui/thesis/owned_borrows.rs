//@rustc-env: CLIPPY_PETS_PRINT=1

struct Animal {
    science_name: String,
    simple_name: String,
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_1(a: String) -> bool {
    a.is_empty()
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_2(a: String) {
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_3(a: String) {
    take_2_loan(&a, &a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_1(mut a: String) {
    a.clear();
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_2(mut a: String) {
    take_1_mut_loan(&mut a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_3(mut a: Animal) {
    take_2_mut_loan(&mut a.science_name, &mut a.simple_name);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed(mut a: String) {
    take_1_mut_loan(&mut a);
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed_2(mut a: Animal) {
    take_2_mixed_loan(&a.science_name, &mut a.simple_name);
}

/// https://github.com/nikomatsakis/nll-rfc/issues/37
#[warn(clippy::borrow_pats)]
fn two_phase_borrow_1(mut vec: Vec<usize>) {
    vec.push(vec.len());
}

#[warn(clippy::borrow_pats)]
fn two_phase_borrow_2(mut num: usize, mut vec: Vec<usize>) {
    vec.push({
        num = vec.len();
        num
    })
}

fn take_1_loan(_: &String) {}
fn take_2_loan(_: &String, _: &String) {}
fn take_1_mut_loan(_: &String) {}
fn take_2_mut_loan(_: &mut String, _: &mut String) {}
fn take_2_mixed_loan(_: &String, _: &mut String) {}

fn main() {}
