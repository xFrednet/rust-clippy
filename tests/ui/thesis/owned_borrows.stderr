FuncReals {
    tcx: PrintPrevent,
    multiverse: {},
}
FuncReals {
    tcx: PrintPrevent,
    multiverse: {},
}
FuncReals {
    tcx: PrintPrevent,
    multiverse: {},
}
thread 'rustc' panicked at clippy_lints/src/borrow_pats/util.rs:170:17:
assertion `left == right` failed
  left: Mut
 right: Not
stack backtrace:
   0:     0x77032cf8caf6 - std::backtrace_rs::backtrace::libunwind::trace::hf298c77755f5b12e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x77032cf8caf6 - std::backtrace_rs::backtrace::trace_unsynchronized::h4da6a489ce2db74e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x77032cf8caf6 - std::sys_common::backtrace::_print_fmt::h051beca4a6f5cb98
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x77032cf8caf6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcb86488900881a45
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x77032cfdda7c - core::fmt::rt::Argument::fmt::haf43875fa4a971bb
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/rt.rs:142:9
   5:     0x77032cfdda7c - core::fmt::write::h44c7f32ae3be5535
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/mod.rs:1120:17
   6:     0x77032cf814bf - std::io::Write::write_fmt::hc3fec0223f64c010
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/io/mod.rs:1846:15
   7:     0x77032cf8c8a4 - std::sys_common::backtrace::_print::h7c03376d42fd745a
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x77032cf8c8a4 - std::sys_common::backtrace::print::h9ff10455f68b06c1
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x77032cf8f5eb - std::panicking::default_hook::{{closure}}::he4f9865f1b72b2d5
  10:     0x77032cf8f339 - std::panicking::default_hook::h3d59f13a05d49fe6
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:292:9
  11:     0x77032fe731ec - std[19d33715048712a6]::panicking::update_hook::<alloc[6fc58ab69ba746fe]::boxed::Box<rustc_driver_impl[4c01069949ad96dd]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x77032cf8fd50 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf41c577e6ac4b0df
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2030:9
  13:     0x77032cf8fd50 - std::panicking::rust_panic_with_hook::h70f59ab226ae2c5c
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:786:13
  14:     0x77032cf8fa92 - std::panicking::begin_panic_handler::{{closure}}::h47fc7126fa960711
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:660:13
  15:     0x77032cf8cfd6 - std::sys_common::backtrace::__rust_end_short_backtrace::h5b94f863278adeaf
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x77032cf8f7c4 - rust_begin_unwind
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:648:5
  17:     0x77032cfd9f95 - core::panicking::panic_fmt::h96a33d9be4ce8c0f
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:72:14
  18:     0x77032cfda4db - core::panicking::assert_failed_inner::hef3141f8e1a6c279
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:342:17
  19:     0x61659f5e279e - core::panicking::assert_failed::ha3cd4ad072e0ea74
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:297:5
  20:     0x61659fbdc4db - clippy_lints::borrow_pats::util::get_parents_of_return::{{closure}}::h1e4d8e9e4bf18c39
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/util.rs:170:17
  21:     0x61659fc2734e - clippy_utils::ty::for_each_ref_region::hb79bd81cdc11d64c
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_utils/src/ty.rs:962:13
  22:     0x61659f5fcac8 - clippy_lints::borrow_pats::util::get_parents_of_return::h71bf884a30fb8aa6
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/util.rs:169:13
  23:     0x61659f5fc072 - clippy_lints::borrow_pats::util::calc_call_local_relations::h5eb7c008b94407c0
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/util.rs:97:20
  24:     0x61659fbdb270 - clippy_lints::borrow_pats::info::meta::MetaAnalysis::visit_terminator_for_terms::h9debf5d1b5e3d2ff
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/info/meta.rs:253:33
  25:     0x61659fbdb8b8 - <clippy_lints::borrow_pats::info::meta::MetaAnalysis as rustc_middle::mir::visit::Visitor>::visit_terminator::he1371170c1fce233
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/info/meta.rs:308:9
  26:     0x61659f7d459f - rustc_middle::mir::visit::Visitor::super_basic_block_data::hbd9e281bcf267909
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:313:21
  27:     0x61659f7d45c7 - rustc_middle::mir::visit::Visitor::visit_basic_block_data::h393412c7b7e1118a
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:88:17
  28:     0x61659f7d29f2 - rustc_middle::mir::visit::Visitor::super_body::h295b5b28655cd86f
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:292:17
  29:     0x61659f7d31b4 - rustc_middle::mir::visit::Visitor::visit_body::h0fae0172fa145739
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:78:17
  30:     0x61659fbd9657 - clippy_lints::borrow_pats::info::meta::MetaAnalysis::from_body::h3c5ad49b7b29edb7
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/info/meta.rs:42:9
  31:     0x61659f7e84ea - clippy_lints::borrow_pats::info::AnalysisInfo::new::he1c9ebaac5a95753
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/info.rs:67:29
  32:     0x61659f689518 - <clippy_lints::borrow_pats::BorrowPats as rustc_lint::passes::LateLintPass>::check_body::hff6ea019ecdb2c22
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/mod.rs:139:28
  33:     0x7703301fa83b - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_body
  34:     0x7703301fb6e4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_fn
  35:     0x7703301eaae4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_item
  36:     0x770331e2d473 - rustc_lint[a04349e93729da6f]::late::check_crate::{closure#0}
  37:     0x770331e2d6d1 - rustc_lint[a04349e93729da6f]::late::check_crate
  38:     0x770331ca39c9 - rustc_interface[3868b2f10de3354]::passes::analysis
  39:     0x770331ca2669 - rustc_query_impl[80d5b36ddf5440ab]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>
  40:     0x770331f33ce5 - rustc_query_system[3c4b6c562f3523cb]::query::plumbing::try_execute_query::<rustc_query_impl[80d5b36ddf5440ab]::DynamicConfig<rustc_query_system[3c4b6c562f3523cb]::query::caches::SingleCache<rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[80d5b36ddf5440ab]::plumbing::QueryCtxt, false>
  41:     0x770331f33a49 - rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  42:     0x770331f42f59 - rustc_interface[3868b2f10de3354]::interface::run_compiler::<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}
  43:     0x7703321bdc45 - std[19d33715048712a6]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>
  44:     0x7703321bda72 - <<std[19d33715048712a6]::thread::Builder>::spawn_unchecked_<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#1} as core[a9f526eb09c6e7cd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  45:     0x77032cf99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h79a9e4aa91d71737
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  46:     0x77032cf99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6eabe626cfdc74a5
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  47:     0x77032cf99725 - std::sys::pal::unix::thread::Thread::new::thread_start::h8428b71c070d7cf2
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys/pal/unix/thread.rs:108:17
  48:     0x770325494ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  49:     0x770325526850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  50:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: please attach the file at `/home/xfrednet/workspace/rust/rust-clippy/rustc-ice-2024-04-18T11_06_05-1173155.txt` to your bug report

note: compiler flags: -Z ui-testing -Z deduplicate-diagnostics=no

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.78 (8079ac210 2024-04-16)

